src/liballoc/sync.rs

```diff
@@ -328,6 +328,70 @@ impl<T> Arc<T> {
         Self::from_inner(Box::leak(x).into())
     }
 
+    /// Constructs a new `Arc<T>` using a weak reference to itself. Attempting
+    /// to upgrade the weak reference before this function returns will result
+    /// in a `None` value. However, the weak reference may be cloned freely and
+    /// stored for use at a later time.
+    ///
+    /// # Examples
+    /// ```
+    /// #![feature(arc_new_with)]
+    /// #![allow(dead_code)]
+    ///
+    /// use std::sync::{Arc, Weak};
+    ///
+    /// struct Foo {
+    ///     me: Weak<Foo>,
+    /// }
+    ///
+    /// let foo = Arc::new_with(|me| Foo {
+    ///     me: me.clone(),
+    /// });
+    /// ```
+    #[inline]
+    #[unstable(feature = "arc_new_with", issue = "none")]
+    pub fn new_with(data_fn: impl FnOnce(&Weak<T>) -> T) -> Arc<T> {
+        // Construct the inner in the "uninitialized" state with a single
+        // weak reference.
+        let uninit_ptr: NonNull<_> = Box::leak(box ArcInner {
+            strong: atomic::AtomicUsize::new(0),
+            weak: atomic::AtomicUsize::new(1),
```

Comment:
Okay so with the old 2 there would have been a leak here? Would be good to have a `#[test]` for the final version as well then; we run those in Miri which means we can tell when there are leaks.

Reply:
Ah never mind you already have such tests. :)
(I checked the force-push diffs but GH skips some of them. :/ )

Reply:
No, when the count was "2" I *was* passing ownership of one of the Weaks so the count was correct. I changed it to pass a reference and start the count at one instead so that if you don't *need* to clone the Weak it doesn't cost anything extra. In between, the comment was wrong as it still said "2" when it should have said "1".

Reply:
Ah, makes sense -- thanks.